cpp/src/arrow/util/hashing.h

```diff
@@ -155,16 +155,21 @@ hash_t ComputeStringHash(const void* data, int64_t length) {
     return n ^ hx ^ hy;
   }
 
-  if (HashUtil::have_hardware_crc32) {
-    // DoubleCrcHash is faster that Murmur2.
-    auto h = HashUtil::DoubleCrcHash(data, static_cast<int32_t>(length), AlgNum);
+  // DoubleCrcHash is faster that Murmur2.
+  if (HashUtil::have_hardware_crc32 == USE_SSE42) {
+    // SSE4.
+    auto h = HashUtil::DoubleHash<USE_SSE42>(data, static_cast<int32_t>(length), AlgNum);
+    return ScalarHelper<uint64_t, AlgNum>::ComputeHash(h);
+  } else if (HashUtil::have_hardware_crc32 == USE_ARMCRC) {
+    // Armce.
+    auto h = HashUtil::DoubleHash<USE_ARMCRC>(data, static_cast<int32_t>(length), AlgNum);
     return ScalarHelper<uint64_t, AlgNum>::ComputeHash(h);
   } else {
     // Fall back on 64-bit Murmur2 for longer strings.
     // It has decent speed for medium-sized strings.  There may be faster
     // hashes on long strings such as xxHash, but that may not matter much
     // for the typical length distribution of hash keys.
-    return HashUtil::MurmurHash2_64(data, static_cast<int>(length), AlgNum);
+    return HashUtil::Hash<USE_DEFAULT>(data, static_cast<int>(length), AlgNum);
```

Comment:
I'd rather keep the explicit fallback to MurmurHash here, rather than `Hash<USE_DEFAULT>`.

Reply:
It's ok.  I'll follow it.